lib/vrl/compiler/src/vm/machine.rs

```diff
@@ -0,0 +1,542 @@
+                    let value = state.read_constant()?;
+                    state.stack.push(value);
+                }
+                OpCode::Not => match state.stack.pop() {
+                    None => return Err("Notting nothing".into()),
+                    Some(Value::Boolean(value)) => state.stack.push(Value::Boolean(!value)),
+                    _ => return Err("Notting non boolean".into()),
+                },
+                OpCode::Add => binary_op(&mut state, Value::try_add)?,
+                OpCode::Subtract => binary_op(&mut state, Value::try_sub)?,
+                OpCode::Multiply => binary_op(&mut state, Value::try_mul)?,
+                OpCode::Divide => binary_op(&mut state, Value::try_div)?,
+                OpCode::Rem => binary_op(&mut state, Value::try_rem)?,
+                OpCode::Merge => binary_op(&mut state, Value::try_merge)?,
+                OpCode::Greater => binary_op(&mut state, Value::try_gt)?,
+                OpCode::GreaterEqual => binary_op(&mut state, Value::try_ge)?,
+                OpCode::Less => binary_op(&mut state, Value::try_lt)?,
+                OpCode::LessEqual => binary_op(&mut state, Value::try_le)?,
+                OpCode::NotEqual => {
+                    let rhs = state.pop_stack()?;
+                    let lhs = state.pop_stack()?;
+                    state.stack.push((!lhs.eq_lossy(&rhs)).into());
+                }
+                OpCode::Equal => {
+                    let rhs = state.pop_stack()?;
+                    let lhs = state.pop_stack()?;
+                    state.stack.push(lhs.eq_lossy(&rhs).into());
+                }
+                OpCode::Pop => {
+                    // Removes the top item from the stack
+                    let _ = state.pop_stack()?;
+                }
+                OpCode::ClearError => {
+                    // Resets the state of the error.
+                    state.error = None;
+                }
+                OpCode::JumpIfFalse => {
+                    // If the value at the top of the stack is false, jump by the given amount
+                    let jump = state.next_primitive()?;
+                    if !is_true(state.peek_stack()?) {
+                        state.instruction_pointer += jump;
+                    }
+                }
+                OpCode::JumpIfTrue => {
+                    // If the value at the top of the stack is true, jump by the given amount
+                    let jump = state.next_primitive()?;
+                    if is_true(state.peek_stack()?) {
+                        state.instruction_pointer += jump;
+                    }
+                }
+                OpCode::JumpIfNotErr => {
+                    // If the current state is not in error, jump by the given amount.
+                    let jump = state.next_primitive()?;
+                    if state.error.is_none() {
+                        state.instruction_pointer += jump;
+                    }
+                }
```

Comment:
The Jump instructions are also used for boolean expressions (And and Or) which don't want the predicate popped.

There's nothing stopping us from creating a new `OpCode` for that if needed, however.

Reply:
Thanks for the reference, interesting! I think that only applies when the language has a concept of "truthy" values, rather than only allowing boolean values in boolean operations. I don't think we support that, do we?

Reply:
We do support other types for OR:

```

$ "nork" || 3
"nork"

$ null || 3
3
```

A lot of our parsing function in particular return objects with fields that can be either string or null, so allowing null coalescing makes it easier than having to have `if null(thing.thang)` everywhere.

Reply:
I see, thanks for clearing up 👍